File: langs/java/util/src/main/java/dev/argon/nobleidl/util/RegenerateApi.java
package dev.argon.nobleidl.util;

import dev.argon.esexpr.KeywordMapping;
import dev.argon.nobleidl.compiler.JavaIDLCompilerOptions;
import dev.argon.nobleidl.compiler.JavaLanguageOptions;
import dev.argon.nobleidl.compiler.JavaNobleIDLCompiler;
import dev.argon.nobleidl.compiler.PackageMapping;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;

public class RegenerateApi {

	public static void main(String[] args) throws Throwable {
		String coreLib = Files.readString(Path.of("../../noble-idl/runtime/nobleidl-core.nidl"));
		String compilerApi = Files.readString(Path.of("../../noble-idl/backend/compiler-api.nidl"));
		String javaAnns = Files.readString(Path.of("../../noble-idl/backend/compiler-api-java-annotations.nidl"));

		JavaNobleIDLCompiler.compile(
			new JavaIDLCompilerOptions(
				Path.of("../runtime/src/gen/java"),
				new JavaLanguageOptions(
					new PackageMapping(
						new KeywordMapping<>(Map.of(
							"nobleidl.core", "dev.argon.nobleidl.runtime"
						))
					)
				),
				List.of(coreLib),
				List.of()
			)
		);

		JavaNobleIDLCompiler.compile(
			new JavaIDLCompilerOptions(
				Path.of("../backend/src/gen/java"),
				new JavaLanguageOptions(
					new PackageMapping(
						new KeywordMapping<>(Map.of(
							"nobleidl.core", "dev.argon.nobleidl.runtime",
							"nobleidl.compiler.api", "dev.argon.nobleidl.compiler.api",
							"nobleidl.compiler.api.java", "dev.argon.nobleidl.compiler.api.java"
						))
					)
				),
				List.of(compilerApi, javaAnns),
				List.of(coreLib)
			)
		);
	}

}

File: langs/scala/util/src/main/scala/nobleidl/util/RegenerateApi.scala
package nobleidl.util

import dev.argon.esexpr.KeywordMapping
import esexpr.Dictionary
import nobleidl.compiler.{PackageImportMapping, ScalaIDLCompilerOptions, ScalaJSLanguageOptions, ScalaLanguageOptions, ScalaNobleIDLCompiler}
import nobleidl.compiler.PackageMapping
import zio.{Scope, ZIO, ZIOAppArgs, ZIOAppDefault}

import scala.jdk.CollectionConverters.*
import java.nio.file.Path

object RegenerateApi extends ZIOAppDefault {

  override def run: ZIO[ZIOAppArgs & Scope, Any, Any] =
    for
      coreLib <- ZIO.readFile(Path.of("../noble-idl/runtime/nobleidl-core.nidl").nn)
      compilerApi <- ZIO.readFile(Path.of("../noble-idl/backend/compiler-api.nidl").nn)
      javaAnns <- ZIO.readFile(Path.of("../noble-idl/backend/compiler-api-java-annotations.nidl").nn)

      _ <- ScalaNobleIDLCompiler.ScalaPlatformRunner.compile(ScalaIDLCompilerOptions(
        outputDir = Path.of("runtime/jvm/src/gen/scala"),
        languageOptions = ScalaLanguageOptions(
          packageMapping = PackageMapping(Dictionary(Map(
            "nobleidl.core" -> "nobleidl.core",
          ))),
          javaAdapters = Some(ScalaLanguageOptions.JavaAdapters(
            packageMapping = PackageMapping(Dictionary(Map(
              "nobleidl.core" -> "dev.argon.nobleidl.runtime",
            ))),
          )),
          jsAdapters = None,
        ),
        inputFileData = Seq(coreLib),
        libraryFileData = Seq(),
      ))

      _ <- ScalaNobleIDLCompiler.ScalaPlatformRunner.compile(ScalaIDLCompilerOptions(
        outputDir = Path.of("runtime/js/src/gen/scala"),
        languageOptions = ScalaLanguageOptions(
          packageMapping = PackageMapping(Dictionary(Map(
            "nobleidl.core" -> "nobleidl.core",
          ))),
          javaAdapters = None,
          jsAdapters = Some(ScalaLanguageOptions.JSAdapters(
            packageMapping = PackageMapping(Dictionary(Map(
              "nobleidl.core" -> "nobleidl.sjs.core",
            ))),
          )),
        ),
        inputFileData = Seq(coreLib),
        libraryFileData = Seq(),
      ))

      _ <- ScalaNobleIDLCompiler.ScalaJSPlatformRunner.compile(ScalaIDLCompilerOptions(
        outputDir = Path.of("runtime/js/src/gen/scala"),
        languageOptions = ScalaJSLanguageOptions(
          packageMapping = PackageMapping(Dictionary(Map(
            "nobleidl.core" -> "nobleidl.sjs.core",
          ))),
          packageImportMapping = PackageImportMapping(Map())
        ),
        inputFileData = Seq(coreLib),
        libraryFileData = Seq(),
      ))

      _ <- ScalaNobleIDLCompiler.ScalaPlatformRunner.compile(ScalaIDLCompilerOptions(
        outputDir = Path.of("backend/src/gen/scala"),
        languageOptions = ScalaLanguageOptions(
          packageMapping = PackageMapping(Dictionary(Map(
            "nobleidl.core" -> "nobleidl.core",
            "nobleidl.compiler.api" -> "nobleidl.compiler.api",
            "nobleidl.compiler.api.java" -> "nobleidl.compiler.api.java",
          ))),
          javaAdapters = Some(ScalaLanguageOptions.JavaAdapters(
            packageMapping = PackageMapping(Dictionary(Map(
              "nobleidl.core" -> "dev.argon.nobleidl.core",
              "nobleidl.compiler.api" -> "dev.argon.nobleidl.compiler.api",
              "nobleidl.compiler.api.java" -> "dev.argon.nobleidl.compiler.api.java",
            )))
          )),
          jsAdapters = None,
        ),
        inputFileData = Seq(compilerApi, javaAnns),
        libraryFileData = Seq(coreLib),
      ))
    yield ()
}
